### PR TITLE
Designs

### PR DESCRIPTION
**Pull request by Urigo**
@helfer thanks for the great idea

@glasser I'll add here some of the discussions we had internally

### CODE REVIEW

File: .designs/authorization.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    * Provides a similar flow to onConnection and onAuthorize with a token, the server can validate and authorize the connection only when the connection created.
+    * The auth token send via a custom WebSocket message.
+    * Features callbacks for `onAuthorize`, `onAuthorized`, `onReject`, `postAuthenticate`.
+* Sockjs-node (https://github.com/sockjs/sockjs-node#authorisation )
+    * Does not provide a built-in solution and suggest a self-implemented authorization.
+* https://gist.github.com/subudeepak/9897212 
+* https://auth0.com/blog/auth-with-socket-io/ 
+* https://auth0.com/blog/auth-with-socket-io/ 
+    * Suggests using authorization on connection using url parameter with the auth token.
+
```
Comment (glasser) [inline .designs/authorization.md]: Worth mentioning what Meteor does.

The DDP protocol itself has no concept of authentication.  Authentication is done via normal method calls.

The Meteor DDP server implementation allows method calls to store state on the connection object representing the current user ID, which is accessible from other methods and publications.  If the user ID ever changes, all publications are basically re-evaluated from scratch (inside the server).  Nothing special is done on the server side to allow methods to notice if the user ID changes while they are running; however, methods run in series unless they explicitly ask to unblock the connection (and the client tries to not send login methods in parallel with other methods).

The Meteor Accounts package tracks DDP connections associated with resume tokens and disconnects them if the resume token associated with that connection is deleted from the database.

Personal opinion: having auth just be "another method" wasn't the best idea.  It works better if it's an established-at-beginning-of-connection, disconnect-to-change thing.  However, the general idea of having a way for changes to authn/authz to "rerun publishers" or "disconnect connections" is nice.

Comment (Urigo) [inline .designs/authorization.md]: mmm I'll merge now anyway, at least so we'll have the folder itself.
we can still keep the discussion here or on related PRs that people do to it

**Comment by Urigo** [comment]
@DxCx feel free to add your comments in a PR

**Comment by sullivanpt** [comment]
I'd really like to be able to pass headers (and origin for that matter) into the Client constructor.

I'm using simple session cookie authentication and I've got the graphql server and the subscriptionsServer listening on the same httpServer so they share cookies. Works in the browser but since I cannot attach cookies to subscriptions-transport-ws Client I cannot send cookies during a mocha integration test.  Letting me pass a headers object into the client Constructor that gets passed to the underlying socket would solve my problem.

On a related note, a client side onSubscribe event is needed for integration tests too; although at least this can be monkey patched.

I'm happy to create a PR for these two additions. Thanks

**Comment by glasser** [comment]
Added my comment in https://github.com/apollostack/graphql-subscriptions/commit/b3a6c8dab6161086f5a48eb4496372833ffb5c19

**Pull request merged**